warehouse/tuf/repository.py

```diff
@@ -0,0 +1,497 @@
+                )
+                role_metadata.sign(SSlibSigner(key), append=True)
+
+            self._store(rolename, role_metadata)
+            snapshot_metadata = self.snapshot_update_meta(
+                rolename, role_metadata.signed.version, snapshot_metadata
+            )
+
+        return snapshot_metadata
+
+    def _check_is_initialized(self) -> None:
+        """
+        Check if Repository metadata is initialized based on the top-level
+        metadata files in the storage backend and update the ``is_initialized``
+        state.
+        """
+        try:
+            if any(
+                role
+                for role in TOP_LEVEL_ROLE_NAMES
+                if isinstance(self.load_role(role), Metadata)
+            ):
+                self._is_initialized = True
+        except StorageError:
+            pass
+
+    def _filename(self, rolename: str, version: int) -> str:
+        """Build the filename based in the role name and version.
+
+        Args:
+            rolename: Role name
+            version: Role version
+
+        Returns:
+             File name
+        """
+        if rolename == Timestamp.type:
+            filename = f"{rolename}.json"
+        else:
+            filename = f"{version}.{rolename}.json"
+
+        return filename
+
+    def _store(self, rolename: str, metadata: Metadata) -> None:
+        """
+        Stores (writes) the Metadata to the Storage using the Storage Backend.
+
+        Args:
+            rolename: Role name
+            metadata: Role Metadata
+        """
+        filename = self._filename(rolename, metadata.signed.version)
+        metadata.to_file(filename, JSONSerializer(), self.storage_backend)
+
+    def _build_delegations(
+        self, rolename: str, delegated_role: DelegatedRole, keys: List[Dict[str, Any]]
+    ) -> Delegations:
+        """
+        Builds the Delegations container object storing keys and DelegatedRole
+        information.
+
+        Args:
+            rolename: Role name
+            delegated_role: Delegated Role
+            keys: List of Keys Dictionary
+
+        Returns
+             The Delegations container object
+             ``tuf.api.metadata.Delegations``
+        """
+        return Delegations(
+            keys={key["keyid"]: Key.from_securesystemslib_key(key) for key in keys},
+            roles={rolename: delegated_role},
+        )
+
+    def initialize(
+        self, payload: Dict[str, RolesPayload], store: Optional[bool]
+    ) -> Dict[str, Metadata]:
+        """
+        Initialize the TUF repository from scratch, including a brand new root.
+
+        Args:
+            payload: Dictionary of the role name and the role parameters
+            store: After initialization stores the metadata files
+
+        Returns:
+             Dictionary with key as role name and value as Metadata
+             ``Dict[str, Metadata]``
+        """
+
+        self._check_is_initialized()
```

Comment:
What's the benefit of a state property `self.is_initialized`, if we don't tust it to be up-to-date?

Reply:
That isn't very clear, I agree.
What do you think about merging all to ``self.is_initialized`` and giving the status in the runtime? Or any suggestion?
example:

```Python
    @property
    def is_initialized(self) -> bool:
        """State of repository

        Returns:
            bool
        """
        try:
            if any(
                role
                for role in TOP_LEVEL_ROLE_NAMES
                if isinstance(self.load_role(role), Metadata)
            ):
                return True
        except StorageError as err:
            if "Can't open" in str(err):
                return True

        return False
```

Reply:
Don't we go EAFP here?  🤔I try to load any TOP_LEVEL_ROLE_NAMES without checking if the files exist?

Reply:
That's true. But I'm thinking about places where you will use this property. 

LYBL:
```python
if is_initialized:
  use_metadata()
else:
  handle_unitialized_repo()

```

EAFP:
```python
try:
  use_metadata()
except StorageError:
  handle_unitialized_repo() # Omit try/except if no need to handle
```

Reply:
OK, I thought about the suggestion I shared.
For example, it wouldn't protect the repository initialization (`initialize()`).